### DISCUSSION TITLE
Is it possible to integrate social logins without requesting userinfo endpoint but instead mapping fields from id token to identity?

### DISCUSSION THREAD

**Discussion by mwxxhdb** [ory/kratos#3362] [Q&A]
Gitlab will return user name, email, avatar url in the ID token, which is enough for use. 
I'd like to know if it is possible to avoid requesting userinfo endpoint?
Can kratos achieve this purpose through configuration?

**Comment by vinckr** [comment]
Hello @mwxxhdb 

Yes, Ory Kratos allows you to map the data returned by the sign-in provider to traits as defined in the identity schema. This is done by creating a Jsonnet code snippet. This snippet becomes a part of the Ory Identities configuration for the given social sign-in provider.

For example, in the case of GitLab, it returns only the sub and sub_legacy claims in the id_token. Ory makes a request to GitLab's /oauth/userinfo API and adds the user info to std.extVar('claims'). However, if GitLab returns the necessary user information in the id_token, you can map these directly to the identity schema.
Here is an example of a Jsonnet code snippet used for data mapping:
```

local claims = {  
 email_verified: false,  
} + std.extVar('claims');  
{  
 identity: {  
 traits: {  
 // Allowing unverified email addresses enables account  
 // enumeration attacks, if the value is used for  
 // verification or as a password login identifier.  
 //  
 // Therefore we only return the email if it (a) exists and (b) is marked verified  
 // by GitLab.  
 [if 'email' in claims && claims.email_verified then 'email' else null]: claims.email,  
 },  
 },  
}
```

In this snippet, the user's email, if it exists and is verified, is mapped to email in the identity schema. You can modify this snippet to map other fields from the id_token to the identity schema as needed.

If the provider includes the necessary information in the id_token, it should be possible to map these directly without needing to request the /userinfo endpoint. However, if the provider does not include this information in the id_token, you might still need to request the /userinfo endpoint.

Let me know if that worked out as expected!
